internal/typeinfo/info.go

```diff
@@ -9,110 +9,182 @@ import (
-
-	// Ordered list of tags
-	tags []string
-
-	tagToField map[string]*structField
+// ArgInfo is used to access type information about SQLair input and output
+// arguments. Methods on ArgInfo can be used to generate input and output value
+// locators.
+//
+// ArgInfo should only be accessed using it methods, not used directly as a
+// map.
+type ArgInfo map[string]arg
+
+// GenerateArgInfo takes sample instantiations of argument types and uses
+// reflection to generate an ArgInfo containing the types.
+func GenerateArgInfo(typeSamples []any) (ArgInfo, error) {
+	argInfo := ArgInfo{}
+	for _, typeSample := range typeSamples {
+		if typeSample == nil {
+			return nil, fmt.Errorf("need struct or map, got nil")
+		}
+		t := reflect.TypeOf(typeSample)
+		switch t.Kind() {
+		case reflect.Struct, reflect.Map:
+			if t.Name() == "" {
+				return nil, fmt.Errorf("cannot use anonymous %s", t.Kind())
+			}
+			info, err := getArgInfo(t)
+			if err != nil {
+				return nil, err
+			}
+			if dupeArg, ok := argInfo[t.Name()]; ok {
+				if dupeArg.typ() == t {
+					return nil, fmt.Errorf("found multiple instances of type %q", t.Name())
+				}
+				return nil, fmt.Errorf("two types found with name %q: %q and %q", t.Name(), dupeArg.typ().String(), t.String())
+			}
+			argInfo[t.Name()] = info
+		case reflect.Pointer:
+			return nil, fmt.Errorf("need struct or map, got pointer to %s", t.Elem().Kind())
+		default:
+			return nil, fmt.Errorf("need struct or map, got %s", t.Kind())
+		}
+	}
+	return argInfo, nil
 }
 
-func (si *structInfo) Typ() reflect.Type {
-	return si.structType
+// InputMember returns an input locator for a member of a struct or map.
+func (argInfo ArgInfo) InputMember(typeName string, memberName string) (Input, error) {
+	vl, err := argInfo.getMember(typeName, memberName)
+	if err != nil {
+		return nil, err
+	}
+	input, ok := vl.(Input)
+	if !ok {
+		return nil, fmt.Errorf("internal error: %s cannot be used as input", vl.ArgType().Kind())
+	}
+	return input, nil
 }
 
-func (si *structInfo) TypeMember(member string) (Member, error) {
-	tm, ok := si.tagToField[member]
+// OutputMember returns an output locator for a member of a struct or map.
+func (argInfo ArgInfo) OutputMember(typeName string, memberName string) (Output, error) {
+	vl, err := argInfo.getMember(typeName, memberName)
+	if err != nil {
+		return nil, err
+	}
+	output, ok := vl.(Output)
 	if !ok {
-		return nil, fmt.Errorf(`type %q has no %q db tag`, si.structType.Name(), member)
+		return nil, fmt.Errorf("internal error: %s cannot be used as output", vl.ArgType().Kind())
 	}
-	return tm, nil
+	return output, nil
 }
 
-func (si *structInfo) GetAllMembers() ([]Member, error) {
+// AllStructOutputs returns a list of output locators that locate every member
+// of the named type along with the names of the members. If the type is not a
+// struct an error is returned.
+func (argInfo ArgInfo) AllStructOutputs(typeName string) ([]Output, []string, error) {
```

Comment:
I love the new name, that addresses all of my concerns in a better way than what I proposed.